server/task.go

```diff
@@ -0,0 +1,277 @@
+package main
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"github.com/arthurfabre/scheduler/api"
+	"github.com/arthurfabre/scheduler/server/pb"
+	"github.com/coreos/etcd/clientv3"
+	"github.com/coreos/etcd/clientv3/clientv3util"
+	"github.com/golang/protobuf/proto"
+	"github.com/satori/go.uuid"
+	"log"
+	"strings"
+	"time"
+)
+
+// Format strings for prefixes. A prefix is a key with everything but the last Task UUID component
+// See README/#ETCD Key Schema
+const (
+	taskPrefix        = "task/"
+	queuedPrefixFmt   = "task/status/queued/"
+	runningPrefixFmt  = "task/status/running/%s/"
+	completePrefixFmt = "task/status/complete/%d/"
+	canceledPrefixFmt = "task/status/canceled/%d/"
+)
+
+// Task handles storing and updating tasks (and their status) in etcd.
+// Tasks embeds the pb.Task implementation generated by protoc.
+type Task struct {
+	*pb.Task
+
+	// Version in etcd when this was last updated / retrieved
+	version int64
+
+	// Key in etcd
+	key string
+}
+
+// watchQueuedTasks returns a Channel of Tasks that have just been queued
+// TODO - Potentially expose DELETE events, so we can backoff before stealing tasks
+func watchQueuedTasks(ctx context.Context, client *clientv3.Client) <-chan *Task {
+	out := make(chan *Task)
+
+	// TODO - Are we cleaning things up properly?
+	go func() {
+		for resp := range client.Watch(ctx, queuedPrefix(), clientv3.WithPrefix(), clientv3.WithFilterDelete()) {
```

Comment:
The documentation on the `clientv3.Watch` states the returned channel is not closed if the context is `context.{Background,TODO}` - this is another signal the context should be "real" to be able to release resources properly (which requires that the context is properly canceled after use).